File: javascript/packages/highlighter/test/diagnostic-renderer.test.ts
import { describe, it, expect, beforeEach } from "vitest"
import dedent from "dedent"

import { themes } from "../src/themes.js"
import { ANSI_REGEX } from "../src/ansi.js"
import { stripAnsiColors } from "./util.js"

import { DiagnosticRenderer } from "../src/diagnostic-renderer.js"
import { SyntaxRenderer } from "../src/syntax-renderer.js"

import type { Diagnostic } from "@herb-tools/core"

describe("DiagnosticRenderer", () => {
  let renderer: DiagnosticRenderer
  let syntaxRenderer: SyntaxRenderer

  beforeEach(async () => {
    syntaxRenderer = new SyntaxRenderer(themes.onedark)
    await syntaxRenderer.initialize()
    renderer = new DiagnosticRenderer(syntaxRenderer)
  })

  const createDiagnostic = (
    overrides: Partial<Diagnostic> = {},
  ): Diagnostic => ({
    message: "Test error message",
    severity: "error",
    location: {
      start: { line: 2, column: 5 },
      end: { line: 2, column: 10 },
    },
    code: "test-rule",
    ...overrides,
  })

  describe("renderSingle", () => {
    it("should render a single error diagnostic", () => {
      const diagnostic = createDiagnostic()
      const content = dedent`
        line 1
        line <error> content
        line 3
      `
      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should render a single warning diagnostic", () => {
      const diagnostic = createDiagnostic({ severity: "warning" })
      const content = dedent`
        line 1
        line <warn> content
        line 3
      `
      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should handle custom context lines", () => {
      const diagnostic = createDiagnostic({
        location: {
          start: { line: 5, column: 1 },
          end: { line: 5, column: 5 },
        },
      })
      const content = dedent`
        line 1
        line 2
        line 3
        line 4
        line 5 error
        line 6
        line 7
      `

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { contextLines: 1 },
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should hide line numbers when requested", () => {
      const diagnostic = createDiagnostic()
      const content = dedent`
        line 1
        line <error> content
        line 3
      `
      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { showLineNumbers: false },
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should handle edge cases for line boundaries", () => {
      const diagnostic = createDiagnostic({
        location: {
          start: { line: 1, column: 1 },
          end: { line: 1, column: 5 },
        },
      })
      const content = "single line"

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { contextLines: 5 },
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should highlight backticks in messages", () => {
      const diagnostic = createDiagnostic({
        message: "Error with `code` in message",
      })
      const content = dedent`
        line 1
        line <error> content
        line 3
      `
      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should handle multi-character error ranges", () => {
      const diagnostic = createDiagnostic({
        location: {
          start: { line: 2, column: 5 },
          end: { line: 2, column: 15 },
        },
      })
      const content = dedent`
        line 1
        line <long error> content
        line 3
      `
      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })
  })

  describe("error handling", () => {
    it("should handle invalid line numbers gracefully", () => {
      const diagnostic = createDiagnostic({
        location: {
          start: { line: 999, column: 1 },
          end: { line: 999, column: 5 },
        },
      })
      const content = dedent`
        line 1
        line 2
      `

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should handle invalid column numbers gracefully", () => {
      const diagnostic = createDiagnostic({
        location: {
          start: { line: 2, column: 999 },
          end: { line: 2, column: 1000 },
        },
      })
      const content = dedent`
        line 1
        short
        line 3
      `

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })
  })

  describe("multi-line diagnostics", () => {
    const multiLineContent = dedent`
      <div>

        <div id="gems">
          <% @gems.each do |topic_gem| %>
            <%= render partial: "gem_card" %>
          <% end %>
        </div>

      </div>
    `

    const multiLineDiagnostic = createDiagnostic({
      message: "Multi-line offense",
      severity: "warning",
      location: {
        start: { line: 4, column: 4 },
        end: { line: 6, column: 13 },
      },
      code: "multi-line-rule",
    })

    it("marks every line the diagnostic spans and expands the trailing context off the end line", () => {
      const result = renderer.renderSingle("/test/file.erb", multiLineDiagnostic, multiLineContent, {
        wrapLines: false,
        contextLines: 2,
      })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("does not mark blank lines inside the span", () => {
      const content = dedent`
        <% if true %>

        <% end %>
      `

      const diagnostic = createDiagnostic({
        location: {
          start: { line: 1, column: 0 },
          end: { line: 3, column: 9 },
        },
      })

      const result = renderer.renderSingle("/test/file.erb", diagnostic, content, { wrapLines: false })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("aligns markers to the content when line numbers are hidden", () => {
      const result = renderer.renderSingle("/test/file.erb", multiLineDiagnostic, multiLineContent, {
        wrapLines: false,
        contextLines: 2,
        showLineNumbers: false,
      })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("keeps a single marker for single-line diagnostics", () => {
      const content = dedent`
        line 1
        line <error> content
        line 3
      `
      const result = renderer.renderSingle("/test/file.erb", createDiagnostic(), content, { wrapLines: false })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })
  })

  describe("smart diagnostic truncation", () => {
    it("should show ellipsis at end when diagnostic is at start of long line", () => {
      const longLineStart = `<div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Class name should be shorter",
        location: {
          start: { line: 1, column: 13 },
          end: { line: 1, column: 33 }
        },
        code: "class-name-length"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        longLineStart,
        {
          truncateLines: true,
          maxWidth: 60
        }
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should show ellipsis at beginning when diagnostic is at end of long line", () => {
      const longLineEnd = `<div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Content should be more descriptive",
        severity: "warning",
        location: {
          start: { line: 1, column: 95 },
          end: { line: 1, column: 102 }
        },
        code: "content-description"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        longLineEnd,
        {
          truncateLines: true,
          maxWidth: 60
        }
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should show ellipsis on both sides when diagnostic is in middle of long line", () => {
      const longLineMiddle = `<div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long-with-more-content">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Avoid 'should-be' in class names",
        location: {
          start: { line: 1, column: 45 },
          end: { line: 1, column: 54 }
        },
        code: "class-naming-convention"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        longLineMiddle,
        {
          truncateLines: true,
          maxWidth: 60
        }
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should adjust pointer position correctly for truncated diagnostics", () => {
      const longLine = `<div class="this-is-a-very-long-class-name-that-should-be-truncated">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Test diagnostic positioning",
        location: {
          start: { line: 1, column: 50 },
          end: { line: 1, column: 55 }
        },
        code: "test-positioning"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        longLine,
        {
          truncateLines: true,
          maxWidth: 40
        }
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should handle truncation with context lines", () => {
      const content = dedent`
        <div class="short-line">Short</div>
        <div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long">Content</div>
        <div class="another-short-line">Short</div>
      `

      const diagnostic = createDiagnostic({
        message: "Long class name detected",
        location: {
          start: { line: 2, column: 13 }, // Points to long class name
          end: { line: 2, column: 30 }
        },
        code: "class-name-length"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        {
          truncateLines: true,
          maxWidth: 60,
          contextLines: 1
        }
      )

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("should not truncate when maxWidth is sufficient", () => {
      const shortLine = `<div class="short">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Test short line",
        location: {
          start: { line: 1, column: 13 },
          end: { line: 1, column: 18 }
        },
        code: "test-short"
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        shortLine,
        {
          truncateLines: true,
          maxWidth: 100
        }
      )

      const strippedResult = stripAnsiColors(result)

      expect(strippedResult).not.toContain("…")
      expect(strippedResult).toMatchSnapshot()
    })
  })

  describe("ANSI-aware truncation preserves styling", () => {
    it("should preserve syntax highlighting colors in truncated output", () => {
      const content = `<div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long">Content</div>`

      const diagnostic = createDiagnostic({
        message: "Line too long",
        location: {
          start: { line: 1, column: 1 },
          end: { line: 1, column: 5 },
        },
        code: "line-length",
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { truncateLines: true, maxWidth: 50 },
      )

      expect(result).toMatch(ANSI_REGEX)
      expect(result).toMatchSnapshot()
    })

    it("should preserve colors when extracting from end of styled line", () => {
      const content = `<div class="this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long">ShortEnd</div>`

      const diagnostic = createDiagnostic({
        message: "End content issue",
        location: {
          start: { line: 1, column: 95 },
          end: { line: 1, column: 103 },
        },
        code: "end-content",
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { truncateLines: true, maxWidth: 50 },
      )

      expect(result).toMatch(ANSI_REGEX)
      expect(result).toMatchSnapshot()
    })

    it("should preserve colors when extracting from middle of styled line", () => {
      const content = `<div class="aaaaaaaaaaaaaaaa-MIDDLE_TARGET-bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb">end</div>`

      const diagnostic = createDiagnostic({
        message: "Middle issue",
        location: {
          start: { line: 1, column: 30 },
          end: { line: 1, column: 44 },
        },
        code: "middle-content",
      })

      const result = renderer.renderSingle(
        "/test/file.erb",
        diagnostic,
        content,
        { truncateLines: true, maxWidth: 50 },
      )

      expect(result).toMatch(ANSI_REGEX)
      expect(result).toMatchSnapshot()
    })
  })

  describe("NO_COLOR environment", () => {
    it("should respect NO_COLOR environment variable", () => {
      const originalNoColor = process.env.NO_COLOR
      process.env.NO_COLOR = "1"

      try {
        const diagnostic = createDiagnostic()
        const content = dedent`
          line 1
          line <error> content
          line 3
        `
        const result = renderer.renderSingle(
          "/test/file.erb",
          diagnostic,
          content,
        )

        expect(result).not.toMatch(ANSI_REGEX)
        expect(result).toMatchSnapshot()
      } finally {
        if (originalNoColor === undefined) {
          delete process.env.NO_COLOR
        } else {
          process.env.NO_COLOR = originalNoColor
        }
      }
    })
  })
})

File: rust/herb-highlighter/tests/diagnostic_renderer_test.rs
mod common;

use herb_highlighter::ansi::find_ansi_sequences;
use herb_highlighter::diagnostic::{Diagnostic, DiagnosticSeverity, SerializedLocation};
use herb_highlighter::diagnostic_renderer::{DiagnosticRenderOptions, DiagnosticRenderer};
use herb_highlighter::syntax_renderer::SyntaxRenderer;
use herb_highlighter::themes::{get_theme, Theme};

use common::{no_color, strip_ansi_colors, with_color};

fn syntax_renderer() -> SyntaxRenderer {
  SyntaxRenderer::new(get_theme(Theme::OneDark).clone())
}

fn diagnostic() -> Diagnostic {
  Diagnostic::new("Test error message", SerializedLocation::from(2, 5, 2, 10), DiagnosticSeverity::Error).with_code("test-rule")
}

#[test]
fn renders_a_single_error_diagnostic() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "line 1\nline <error> content\nline 3";
  let result = renderer.render_single("/test/file.erb", &diagnostic(), content, &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn renders_a_single_warning_diagnostic() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.severity = DiagnosticSeverity::Warning;

  let content = "line 1\nline <warn> content\nline 3";
  let result = renderer.render_single("/test/file.erb", &diagnostic, content, &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_custom_context_lines() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(5, 1, 5, 5);

  let content = "line 1\nline 2\nline 3\nline 4\nline 5 error\nline 6\nline 7";

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      context_lines: 1,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn hides_line_numbers_when_requested() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "line 1\nline <error> content\nline 3";

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic(),
    content,
    &DiagnosticRenderOptions {
      show_line_numbers: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_edge_cases_for_line_boundaries() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(1, 1, 1, 5);

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    "single line",
    &DiagnosticRenderOptions {
      context_lines: 5,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn highlights_backticks_in_messages() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.message = "Error with `code` in message".to_string();

  let content = "line 1\nline <error> content\nline 3";
  let result = renderer.render_single("/test/file.erb", &diagnostic, content, &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_multi_character_error_ranges() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(2, 5, 2, 15);

  let content = "line 1\nline <long error> content\nline 3";
  let result = renderer.render_single("/test/file.erb", &diagnostic, content, &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_invalid_line_numbers_gracefully() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(999, 1, 999, 5);

  let result = renderer.render_single("/test/file.erb", &diagnostic, "line 1\nline 2", &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_invalid_column_numbers_gracefully() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(2, 999, 2, 1000);

  let result = renderer.render_single("/test/file.erb", &diagnostic, "line 1\nshort\nline 3", &DiagnosticRenderOptions::default());

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

const MULTI_LINE_CONTENT: &str =
  "<div>\n\n  <div id=\"gems\">\n    <% @gems.each do |topic_gem| %>\n      <%= render partial: \"gem_card\" %>\n    <% end %>\n  </div>\n\n</div>";

fn multi_line_diagnostic() -> Diagnostic {
  Diagnostic::new("Multi-line offense", SerializedLocation::from(4, 4, 6, 13), DiagnosticSeverity::Warning).with_code("multi-line-rule")
}

#[test]
fn marks_every_line_the_diagnostic_spans_and_expands_the_trailing_context() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let result = renderer.render_single(
    "/test/file.erb",
    &multi_line_diagnostic(),
    MULTI_LINE_CONTENT,
    &DiagnosticRenderOptions {
      wrap_lines: false,
      context_lines: 2,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn does_not_mark_blank_lines_inside_the_span() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let mut diagnostic = diagnostic();
  diagnostic.location = SerializedLocation::from(1, 0, 3, 9);

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    "<% if true %>\n\n<% end %>",
    &DiagnosticRenderOptions {
      wrap_lines: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn aligns_markers_to_the_content_when_line_numbers_are_hidden() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let result = renderer.render_single(
    "/test/file.erb",
    &multi_line_diagnostic(),
    MULTI_LINE_CONTENT,
    &DiagnosticRenderOptions {
      wrap_lines: false,
      context_lines: 2,
      show_line_numbers: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn keeps_a_single_marker_for_single_line_diagnostics() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic(),
    "line 1\nline <error> content\nline 3",
    &DiagnosticRenderOptions {
      wrap_lines: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn shows_ellipsis_at_end_when_diagnostic_is_at_start_of_long_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long\">Content</div>";

  let diagnostic = Diagnostic::new(
    "Class name should be shorter",
    SerializedLocation::from(1, 13, 1, 33),
    DiagnosticSeverity::Error,
  )
  .with_code("class-name-length");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(60),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn shows_ellipsis_at_beginning_when_diagnostic_is_at_end_of_long_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long\">Content</div>";

  let diagnostic = Diagnostic::new(
    "Content should be more descriptive",
    SerializedLocation::from(1, 95, 1, 102),
    DiagnosticSeverity::Warning,
  )
  .with_code("content-description");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(60),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn shows_ellipsis_on_both_sides_when_diagnostic_is_in_middle_of_long_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long-with-more-content\">Content</div>";

  let diagnostic = Diagnostic::new(
    "Avoid 'should-be' in class names",
    SerializedLocation::from(1, 45, 1, 54),
    DiagnosticSeverity::Error,
  )
  .with_code("class-naming-convention");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(60),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn adjusts_pointer_position_correctly_for_truncated_diagnostics() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated\">Content</div>";

  let diagnostic =
    Diagnostic::new("Test diagnostic positioning", SerializedLocation::from(1, 50, 1, 55), DiagnosticSeverity::Error).with_code("test-positioning");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(40),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn handles_truncation_with_context_lines() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"short-line\">Short</div>\n<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long\">Content</div>\n<div class=\"another-short-line\">Short</div>";

  let diagnostic =
    Diagnostic::new("Long class name detected", SerializedLocation::from(2, 13, 2, 30), DiagnosticSeverity::Error).with_code("class-name-length");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(60),
      context_lines: 1,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn does_not_truncate_when_max_width_is_sufficient() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let diagnostic = Diagnostic::new("Test short line", SerializedLocation::from(1, 13, 1, 18), DiagnosticSeverity::Error).with_code("test-short");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    "<div class=\"short\">Content</div>",
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(100),
      ..Default::default()
    },
  );

  let stripped = strip_ansi_colors(&result);

  assert!(!stripped.contains('…'));

  insta::assert_snapshot!(stripped);
}

#[test]
fn preserves_syntax_highlighting_colors_in_truncated_output() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long\">Content</div>";

  let diagnostic = Diagnostic::new("Line too long", SerializedLocation::from(1, 1, 1, 5), DiagnosticSeverity::Error).with_code("line-length");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(50),
      ..Default::default()
    },
  );

  assert!(!find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

#[test]
fn preserves_colors_when_extracting_from_end_of_styled_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"this-is-a-very-long-class-name-that-should-be-truncated-when-the-line-is-too-long\">ShortEnd</div>";

  let diagnostic = Diagnostic::new("End content issue", SerializedLocation::from(1, 95, 1, 103), DiagnosticSeverity::Error).with_code("end-content");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(50),
      ..Default::default()
    },
  );

  assert!(!find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

#[test]
fn preserves_colors_when_extracting_from_middle_of_styled_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "<div class=\"aaaaaaaaaaaaaaaa-MIDDLE_TARGET-bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb\">end</div>";

  let diagnostic = Diagnostic::new("Middle issue", SerializedLocation::from(1, 30, 1, 44), DiagnosticSeverity::Error).with_code("middle-content");

  let result = renderer.render_single(
    "/test/file.erb",
    &diagnostic,
    content,
    &DiagnosticRenderOptions {
      truncate_lines: true,
      max_width: Some(50),
      ..Default::default()
    },
  );

  assert!(!find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

#[test]
fn respects_the_no_color_setting() {
  no_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiagnosticRenderer::new(&syntax_renderer);

  let content = "line 1\nline <error> content\nline 3";
  let result = renderer.render_single("/test/file.erb", &diagnostic(), content, &DiagnosticRenderOptions::default());

  assert!(find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}
